### PR TITLE
Prevent memory leaks in StreamLogFiles

### PR DESCRIPTION
**Pull request by xnull**
## Overview
Possible memory leaks in StreamLogFiles

Description:
memory leaks could happen in the current version of StreamLogFiles

Why should this be merged: 
Prevent memory leaks in the stream log layer.

There are methods:
- `SegmentHandle.close()` -> `channel.force(true);` can throw an exception and there is no finally construction to close the resource. The channel will remain unclosed.

- `StreamLogFiles.trimLogFile()->trimLogFile->getCompactedEntries->getChannel` - a file channel can be opened and then an exception can be thrown, so the channel remains open.

- `StreamLogFiles.getSegmentHandleForAddress()` - we create 4 channels `writeCh, readCh, trimmedCh, pendingTrimmedCh` we can create all these channels and then throw RuntimeException (see `getChannel` method) in the result all these channels ctay opened.

 - `verifyLogs()` - we excute `fc.close();` inside `try` section. If an exception happens in a previous line (which is `parseHeader(fc);`) there could be a leak. But we get the FileChannel from the FileInputStream, there should NOT be a resource leak. I just simplified that piece of code to make it clear and do not confuse a developer who could read it.

Related issue(s) (if applicable): #1474

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 1]
```diff
@@ -57,19 +58,19 @@ public synchronized void release() {
...
+
+            try {
                 channel.close();
-                channel = null;
             } catch (Exception e) {
                 log.warn("Error closing channel {}: {}", channel.toString(), e.toString());
             }
         }
-
-        knownAddresses = null;
```
Comment (StrikingThirteen) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: There's a minor change of behavior here: as with making these fields final, the corresponding references are not being released by close() anymore.

Comment (xnull) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: In addition: https://www.ibm.com/developerworks/java/library/j-jtp01274/index.html#3.2
Especially section: Immutability, again

**Comment by xnull** [comment]
There are methods:
- `SegmentHandle.close()` - `channel.force(true);` can throw an exception and there is no finally construction to close the resource. The channel will remain unclosed.

- `StreamLogFiles.trimLogFile()->trimLogFile->getCompactedEntries->getChannel` - a file channel can be opened and then an exception can be thrown, so the channel remains open.

- `StreamLogFiles.getSegmentHandleForAddress()` - we create 4 channels `writeCh, readCh, trimmedCh, pendingTrimmedCh` we can create all these channels and then throw RuntimeException (see `getChannel` method) in the result all these channels ctay opened.

 - `verifyLogs()` - we excute `fc.close();` inside `try` section. If an exception happens in a previous line (which is `parseHeader(fc);`) there could be a leak. But we get the FileChannel from the FileInputStream, there should NOT be a resource leak. I just simplified that piece of code to make it clear and do not confuse a developer who could read it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 2]
```diff
@@ -57,19 +58,19 @@ public synchronized void release() {
 
     public void close() {
         Set<FileChannel> channels =
-                new HashSet(Arrays.asList(writeChannel, readChannel, trimmedChannel, pendingTrimChannel));
+                new HashSet<>(Arrays.asList(writeChannel, readChannel, trimmedChannel, pendingTrimChannel));
         for (FileChannel channel : channels) {
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: Why wrap force with a nested try/catch if you can just add a finally clause?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 3]
```diff
@@ -787,9 +784,11 @@ private LogData  readRecord(SegmentHandle sh, long address)
...
 
                 try {
-                    FileChannel channel = FileChannel.open(FileSystems.getDefault().getPath(filePath),
-                            EnumSet.of(StandardOpenOption.READ, StandardOpenOption.WRITE,
-                                    StandardOpenOption.CREATE_NEW, StandardOpenOption.SPARSE));
+                    EnumSet<StandardOpenOption> options = EnumSet.of(
+                            StandardOpenOption.READ, StandardOpenOption.WRITE,
+                            StandardOpenOption.CREATE_NEW, StandardOpenOption.SPARSE
+                    );
+                    FileChannel channel = FileChannel.open(FileSystems.getDefault().getPath(filePath), options);
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: BLOCKER Use try-with-resources or close this "FileChannel" in a "finally" clause. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2095)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 1 issue
* ![BLOCKER][BLOCKER] 1 blocker

Watch the comments in this conversation to review them.

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'

**Pull request merged**